lib/TPP/Dialect/Tpp/TppUtils.cpp

```diff
@@ -138,106 +135,105 @@ bool hasCopySemantics(linalg::LinalgOp linalgOp) {
+static bool isZeroValue(Value val, Operation *userOp) {
+  if (!val)
+    return false;
+  if (isValConstZero(val))
+    return true;
 
-    ++userIt;
+  Operation *defOp = nullptr;
+
+  // Block arguments don't have a defining op, but they do have an op arg
+  if (auto arg = dyn_cast<BlockArgument>(val)) {
+    // We need to find the argument to the linalg on the same order as this one
+    auto *linalgOp = arg.getParentRegion()->getParentOp();
+    if (!isa<linalg::GenericOp>(linalgOp))
+        return false;
+    auto index = arg.getArgNumber();
+    auto linalgArg = linalgOp->getOperand(index);
+    defOp = linalgArg.getDefiningOp();
+  } else {
+    defOp = val.getDefiningOp();
   }
 
-  return nullptr;
+  return isZeroOp(defOp, userOp);
 }
 
-bool isValConstZero(Value val) {
-  return matchPattern(val, m_AnyZeroFloat()) || matchPattern(val, m_Zero());
-}
+// This is a wrapper to isZeroValue as the entry point of the recursive pattern
+bool isZeroTensor(Value val) { return isZeroValue(val, nullptr); }
 
-bool isZeroTensor(Value val) { return isZeroTensor(val.getDefiningOp()); }
+// Returns true if the attribute represent "all zeros"
+bool isZeroAttr(Attribute attribute) {
+  return TypeSwitch<Attribute, bool>(attribute)
+      .Case<FloatAttr>([](auto attr) {
+          return attr.getValueAsDouble() == 0.0;
+      })
+      .Case<IntegerAttr>([](auto attr) {
+          return attr.getInt() == 0;
+      })
+      .Case<DenseElementsAttr>([](auto attr) {
+        if (!attr.getElementType().isIntOrFloat())
+          return false;
+        if (!attr.isSplat())
+          return false;
+        auto value = attr.template getSplatValue<Attribute>();
+        if (auto floatAttr = dyn_cast<FloatAttr>(value)) {
+          if (floatAttr.getValueAsDouble() == 0.0)
+            return true;
+        }
+        return false;
+      })
+      .Default([](auto attr) { return false; });
+}
 
 // Returns true if the operation represents a zero filled tensor
-static bool isZeroTensor(Operation *defOp) {
+// Recurses into isZeroValue for operands and isZeroAttr for attributes
+static bool isZeroOp(Operation *defOp, Operation *userOp) {
   if (!defOp)
     return false;
 
-  // TODO: add more possible sources of zero filled tensors
-  // TODO: propagate operands of other operations that do not modify underlying
-  //       data values
   return TypeSwitch<Operation *, bool>(defOp)
+      .Case<arith::ConstantOp>([&](auto op) {
+        // Dense attributes don't match APFloat.isZero()
+        auto attr = op.getValue();
+        attr.dump();
+        return isZeroAttr(attr);
+      })
       .Case<linalg::FillOp>([&](auto op) {
-        auto inputs = op.getInputs();
-        return inputs.size() == 1 ? isValConstZero(inputs[0]) : false;
+        if (op.getInputs().size() != 1)
+          return false;
+        return isZeroValue(op.getInputs()[0], op);
       })
       .Case<linalg::CopyOp>([&](auto op) {
```

Comment:
`FillOp` and `CopyOp` cases can be combined.

Reply:
Done